compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -716,6 +716,82 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
 }
 
 impl EncodeContext<'a, 'tcx> {
+    fn encode_def_ids(&mut self) {
+        if self.is_proc_macro {
+            return;
+        }
+        let tcx = self.tcx;
+        let hir = tcx.hir();
+        for local_id in hir.iter_local_def_id() {
+            let def_id = local_id.to_def_id();
+            let def_kind = tcx.opt_def_kind(local_id);
+            let def_kind = if let Some(def_kind) = def_kind { def_kind } else { continue };
+            record!(self.tables.def_kind[def_id] <- match def_kind {
+                // Replace Ctor by the enclosing object to avoid leaking details in children crates.
+                DefKind::Ctor(CtorOf::Struct, _) => DefKind::Struct,
+                DefKind::Ctor(CtorOf::Variant, _) => DefKind::Variant,
+                def_kind => def_kind,
+            });
+            record!(self.tables.span[def_id] <- tcx.def_span(def_id));
+            record!(self.tables.attributes[def_id] <- tcx.get_attrs(def_id));
+            record!(self.tables.expn_that_defined[def_id] <- self.tcx.expansion_that_defined(def_id));
+            if let DefKind::Mod
+            | DefKind::Struct
+            | DefKind::Union
+            | DefKind::Enum
+            | DefKind::Variant
+            | DefKind::Trait
+            | DefKind::TyAlias
+            | DefKind::ForeignTy
+            | DefKind::TraitAlias
+            | DefKind::AssocTy
+            | DefKind::Fn
+            | DefKind::Const
+            | DefKind::Static
+            | DefKind::Ctor(..)
+            | DefKind::AssocFn
+            | DefKind::AssocConst
+            | DefKind::Macro(..)
+            | DefKind::Use
+            | DefKind::ForeignMod
+            | DefKind::OpaqueTy
+            | DefKind::Impl
+            | DefKind::Field = def_kind
+            {
+                record!(self.tables.visibility[def_id] <- self.tcx.visibility(def_id));
```

Comment:
One possibility would be to add an intermediate `visibilities` query which is just a map from local def ids to visibilities and have the `visibility` query fetch its information from that instead of from the HIR directly. Then we could iterate that `visibilities` query result here instead of invoking `visibility` on a certain set of visibilities. That would duplicate all visibility entries though, so it's not ideal either. 

For the reduce of complexity, I'm fine merging the PR as it is and taking the up to 1% perf regression

Reply:
I don't understand. Isn't that what the visibility query does already? It fetches the visibilities computed by resolving, and corrects the missing elements. I can investigate doing this eagerly during HIR indexing.

If you are ok with the large `if let`, I will remove the `visibility -> try_visibility` rename before merging.

Reply:
Oh... I guess then you can remove the visibiliy query check entirely and just iterate the `tcx.visibilites` (instead of all def ids) to find out which visibility queries to run? Not sure if that is too hacky, maybe we should uplift `tcx.visibilities` to a query that just returns the field... cc @petrochenkov

Reply:
The `tcx.visibilities` table is incomplete, it's not enough to iterate it to encode all visibilities.
In theory is can be completed during AST lowering, but filling the missing entries on the fly in the query was simpler.
If the table is completed during AST lowering, then iterating it will be enough to encode all visibilities.